openerp/addons/base/ir/ir_translation.py

```diff
@@ -354,28 +377,134 @@ def _get_source(self, cr, uid, name, types, lang, source=None, res_id=None):
                 res_id = tuple(res_id)
         return self.__get_source(cr, uid, name, types, lang, source, res_id)
 
-    def create(self, cr, uid, vals, context=None):
-        if context is None:
-            context = {}
-        ids = super(ir_translation, self).create(cr, uid, vals, context=context)
-        self.__get_source.clear_cache(self)
-        self._get_ids.clear_cache(self)
-        self.pool['ir.ui.view'].clear_cache()
-        return ids
+    @api.model
+    def _get_terms_query(self, model_name, field_name, lang, ids):
+        """ Utility function that makes the query for field terms. """
+        query = """ SELECT id, res_id, src, value FROM ir_translation
+                    WHERE lang=%s AND type=%s AND name=%s AND res_id IN %s """
+        params = (lang, 'model', "%s,%s" % (model_name, field_name), tuple(ids))
+        return query, params
+
+    @api.model
+    def _get_terms_mapping(self, model_name, field_name, lang, ids):
+        """ Return a function mapping a (translation_id, value) to a value. """
+        if self._context.get('edit_translations'):
+            return lambda tid, value: '<span data-oe-translation-id="%s">%s</span>' % (tid, value)
+        else:
+            return lambda _, value: value
 
-    def write(self, cursor, user, ids, vals, context=None):
-        if context is None:
-            context = {}
-        if isinstance(ids, (int, long)):
-            ids = [ids]
-        if vals.get('src') or ('value' in vals and not(vals.get('value'))):
-            vals.update({'state':'to_translate'})
+    @api.model
+    def _get_terms_translations(self, model_name, field_name, lang, ids):
+        """ Return the terms and translations of a given field on record ids.
+
+        :return: {record_id: {source: value}}
+        """
+        result = {rid: {} for rid in ids}
+        if ids:
+            query, params = self._get_terms_query(model_name, field_name, lang, ids)
+            self._cr.execute(query, params)
+            map_trans = self._get_terms_mapping(model_name, field_name, lang, ids)
+            for trans_id, res_id, src, value in self._cr.fetchall():
```

Comment:
if `self._context.get('edit_translations')`, we should insert missing translations before retrieving them